lib/Dialect/BaseModelica/Transforms/CallCSE.cpp

```diff
@@ -117,105 +261,223 @@ mlir::Operation *cloneDefUseChain(mlir::Operation *op,
         for (mlir::Value operand : childOp->getOperands()) {
           // If an operand is defined in the same scope as `current`,
           // i.e. the equation template scope, add it to the worklist.
-          mlir::Operation *definingOp = operand.getDefiningOp();
-          if (definingOp && definingOp->getBlock() == current->getBlock()) {
+          if (mlir::Operation *definingOp = operand.getDefiningOp();
+              definingOp && definingOp->getBlock() == current->getBlock()) {
             worklist.push_back(definingOp);
           }
         }
       });
     }
   }
 
-  mlir::IRMapping mapping;
   mlir::Operation *root = nullptr;
   for (mlir::Operation *opToClone : llvm::reverse(toClone)) {
     // Skip repeated dependencies on the same operation
-    if (mapping.contains(opToClone)) {
+    if (inductionMapping.contains(opToClone)) {
       continue;
     }
-    root = rewriter.clone(*opToClone, mapping);
+    root = rewriter.clone(*opToClone, inductionMapping);
   }
   return root;
 }
 
-void CallCSEPass::emitCse(llvm::SmallVectorImpl<CallOp> &equivalenceGroup,
-                          ModelOp modelOp, DynamicOp dynamicOp,
-                          mlir::SymbolTable &symbolTable,
-                          mlir::RewriterBase &rewriter) {
-  assert(!equivalenceGroup.empty() && "equivalenceGroup cannot be empty");
-  CallOp representative = equivalenceGroup.front();
-  const mlir::Location loc = representative.getLoc();
+/// Build the template needed to drive [cseVariable] with the result of
+/// [representative].
+EquationTemplateOp
+emitDriverEquationTemplate(const CallEquivalenceGroupEntry &representative,
+                           VariableOp cseVariable, const size_t resultIndex,
+                           mlir::RewriterBase &rewriter, mlir::Location loc) {
+  auto cseEquationTemplateOp = rewriter.create<EquationTemplateOp>(loc);
+  // Create the body of the template with |inductions| = the number of
+  // inductions used by the representative.
+  size_t numInductions = representative.orderedUsedInductions.size();
+  rewriter.setInsertionPointToStart(
+      cseEquationTemplateOp.createBody(numInductions));
+
+  // Build equation left hand side by getting the cse variable reference.
+  auto cseVarRef = rewriter.create<VariableGetOp>(loc, cseVariable);
+  EquationSideOp lhsOp;
+  if (numInductions == 0) {
+    lhsOp = rewriter.create<EquationSideOp>(loc, cseVarRef->getResults());
+  } else {
+    // Extract cse tensor slot with template inductions.
```

Comment:
I'm not sure this is correct.
Let's say the original equation range was [5,8]. The CSE variable will have 4 elements. The driving equation instance will have range [5,8], and will be the only usage of the driving template. Within the template, the accesses to the CSE variable should be shifted back by 5 positions, not 1. Or am I missing something? The other usages of the inductions (i.e., within the function call) should instead remain untouched.

Reply:
I'll think about it for a second, but I think you're right.

Reply:
I fixed it now, and added a test.